### DISCUSSION TITLE
Installing an Eclipse plugin via oomph 1/ doesn't work and 2/ breaks Eclipse installation

### DISCUSSION THREAD

**Discussion by cristian-spiescu** [eclipse-oomph/oomph#94] [Q&A]
Hello,

I am trying to install the plugin [Enhanced Class Decompiler](https://ecd-plugin.github.io/ecd/) in Eclipse via oomph.

The file `.../.eclipse/org.eclipse.oomph.setup/setups/user.setup` looks like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<setup:User
    xmi:version="2.0"
    xmlns:xmi="http://www.omg.org/XMI"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:setup="http://www.eclipse.org/oomph/setup/1.0"
    xmlns:setup.p2="http://www.eclipse.org/oomph/setup/p2/1.0"
    name="poweruser">
  <setupTask
      xsi:type="setup.p2:P2Task">
    <requirement
        name="org.sf.feeling.decompiler.feature.group"/>
    <requirement
        name="org.sf.feeling.decompiler.cfr.feature.group"/>
    <requirement
        name="org.sf.feeling.decompiler.fernflower.feature.group"/>
    <requirement
        name="org.sf.feeling.decompiler.jd.feature.group"/>
    <requirement
        name="org.sf.feeling.decompiler.procyon.feature.group"/>
    <requirement
        name="org.sf.feeling.decompiler.vineflower.feature.group"/>
    <repository
        url="https://ecd-plugin.github.io/update/"/>
  </setupTask>
  <attributeRule
      attributeURI="http://www.eclipse.org/oomph/setup/1.0#//InstallationTask/location"
      value="${install.root/}${installation.id}"/>
  <acceptedLicense>DEFAULT_LICENSE Marker to remember license pre-prompting</acceptedLicense>
  <acceptedLicense>8d3137b6d090b6860d1b977a88ceb334 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>d77cfd8b73fc71776727fcbb2605709 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>6a3d083ad2bd7d3a80ee293235f8c5b1 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>abc76a6cc9d06e4684ff61ed74a972c Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>be4a7d94b4e1903e628a3001859739a8 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>318bcab4617b336391ac7ab40514ccc3 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>aa1a84ed6378c1cdba13a100c225aa57 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>425529a69afe7129289ce9a30b4d2b6c Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>84f0b7f0a433c033764dd28540d6d199 Eclipse Foundation Software User Agreement</acceptedLicense>
  <acceptedLicense>4450aa10f21d146d6a6a80f84275f5b2 Eclipse Foundation Software User Agreement</acceptedLicense>
</setup:User>
```

I edited it by performing drag and drop from the `Repository Explorer`:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/9edb7107-4b55-40cc-b434-52bd28670f1a)

## Results

1/ During the installation I observe this error message that doesn't seem to block the install process:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/b8c1d39e-1ccb-4c6c-889b-9066a53c99b0)

2/ The installation takes a lot of time. It seems to download/install a lot (the majority?) of plugins that already exist in Eclipse. 

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/3428d311-7812-42c9-ac19-6e032b21da0c)

3/ After finishing install/restart, Eclipse seems to be corrupt. Probably because of the plugins that were replaced during 2/:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/07aee95d-78a9-4156-910b-8d33fb69fba5)

Thanks in advance 🙂.
Cristian

**Comment by merks** [comment]
I try this too:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/c6010c1e-8e1c-40fe-98d5-8847d0acb276)

No warning about the bundle pool being broken:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/36bf8e32-93e6-4484-b1b3-3854cfde1716)

Asks me if I trust stuff from there:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/40b92084-6c26-4dc1-886c-848cb16105e7)

It wants me to trust their certificate:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/32ae8796-2f6b-4ff2-9794-479c79795402)

And after a restart it's installed:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/433f4a10-c2f8-4188-9a0c-a0ed490fb414)


It really looks like your shared bundle pool was corrupted.  :scream:  But I'm not sure how that could have happened.   An overly aggressive windows virus scanner deleting the jars?   There's not much you can do to recover from that other than starting from scratch.  😭

**Reply by cristian-spiescu** [reply]
Thanks Ed for your quick reply.

> It really looks like your shared bundle pool was corrupted.

I am using a freshly downloaded eclipse 2023-12. So the eclipse install didn t come from an installer that specified a bundle pool. I am passing an explicit HOME in eclipse.ini. So the question: why is it trying to look in the home dir for p2 stuff? Is this normal?

> aggressive windows virus scanner

I'm on WSL. So no AV.

**Reply by merks** [reply]
Even when I unzipped a packaged zip, it all still works correctly like above and then bundles end up installed local to the installation rather than in a shared pool.

![image](https://github.com/eclipse-oomph/oomph/assets/208716/16ae5cb7-0c5a-4dbe-b256-6d2dea266010)

And the pool is just the local installation itself:

![image](https://github.com/eclipse-oomph/oomph/assets/208716/4f39e1bd-cee7-4529-96db-4f913d1ed637)

Can you clarify with what you mean by "passing an explicit HOME in eclipse.ini"?

**Comment by cristian-spiescu** [comment]
> Can you clarify with what you mean by "passing an explicit HOME in eclipse.ini"?

I mean using in `eclipse.ini`: `-Duser.home=...`.

I come with new findings:

**1/** I tried the above setup w/ Eclipse 2024-06 (w/o using `-Duser.home`, and deleting `~.eclipse`). At first glance it seems to work. However, for me it's a no go, because the plugin I'm installing (decompiler) is not compatible w/ Eclipse 2024-06. Hence I'm forced to use an older Eclipse.

However, during this installation (which was successful), I got the message:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/5712b545-5ba0-4615-88be-475e021f535a)

It's similar to the first picture in my first post, but now w/ more entries.

Where does Eclipse store such information? These directories are directories where I performed various tests.

**2/** I tried to redo the steps w/ the older Eclipse = 2023-12. I got a similar message to the one above. But additionally, the installer would hang:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/5535014c-cfed-48c9-bb99-d7c958fac907)

So again, why in the world is trying Eclipse to look in those directories?

**3/** I searched my home dir, I found that these things seem to come from: ~/.eclipse/org.eclipse.oomph.p2/agents.info`:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/bee2568a-9de2-431b-a7f9-20a3ed3730ed)

I saw that if I remove `~/.eclipse`, I unpack a new Eclipse, I start it: it would repopulate this file. So the info is also stored somewhere else (or maybe the disk is scanned?). 

I removed all the lines except the last one. The result is that during the install I didn't get any more the error w/ "no repository found ...".

However, again, a lot (all?) the plugins reinstalled:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/55a49a85-1f43-4701-8c36-f54a54451a97)

And after restart, the install was corrupt.

Until here, I'd say that this buggy feature of "reinstalling all the plugins of eclipse" happen only in 2023-12, and not in 2024-06 (or -09 cf. your screenshots).

**4/** I tried on Windows, using 2023-12 as well. I got the "no repository found ..." error also:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/e7b27ce7-d51b-442e-844b-84edd2233a76)

And the behaviour "reinstalling all the plugins of eclipse", happened as well. 

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/b6cfd828-40bb-428d-9607-a8a0b7289676)

So I'd say that if you use the package [jee 2023-12](https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/2023-12/R/eclipse-jee-2023-12-R-win32-x86_64.zip), you'll see the issue as well.

**4.1** I'd speculate that during the install process, although I have Eclipse 2023-12, an I request "please install plugin Enhanced Java Decompiler", the program also updates Eclipse by using plugins from Eclipse 2024-06:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/081d9dc7-5983-41e1-8569-26ba4b3a6453)

So this update of Eclipse:
1/ was not requested/wanted.
2/ is a thing at which, in my opinion, Eclipse was never good at. In 20 years of using Eclipse, I never managed to easily and seamlessly upgrade it. I always fell back to "use a new eclipse dir".

**Comment by merks** [comment]
In the Install/Update preferences for Available update sites do you have lots of sites listed?  This list tends to grow and will be used by the p2 task. Note too in the home folder both a .p2 and a .eclipse folder are created.  The former will contain the shared pool by default. 

In my pictures you can see that I too set user.home system property in the eclipse.ini so I can test in a clean new environment each time I need to do that. 

If you run the installer with the arguments 
> -vmargs 
-Duser.home=xxx -Doomph.setup.user.home.redirect=true

It will behave as if that is the user home and will add that -Duser.home to the installation. That’s how I do my testing. 

In advanced mode the installer can install older versions of Eclipse.

**Reply by cristian-spiescu** [reply]
> do you have lots of sites listed?

When I just start w/ a new Eclipse, the list is rather small:

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/a2ab817b-15f1-4135-bc9f-4433ccf7b2a1)

But after the install process starts, the list becomes bigger. And based by the name of some entries, I'd say they are the source for 2024-06 packages (although my installed Eclipse is 2023-12).

![image](https://github.com/eclipse-oomph/oomph/assets/4469897/4e8e9536-a373-419f-8ec3-8c5fb04b6b55)

Why does this happen? At least one update site points to the latest Eclipse, isn't it? Judging by name/URL, maybe one of the 2 entries w/ "Latest Eclipse ...".

I have unchecked all of them. And I confirm that now the install is working correctly.

But how can I automatize this operation (the fact that I unchecked)? 
* Is there some flag that would prevent using them?
* I saw that "preferences recorder" doesn't detect the unchecking. One idea would have been to go this way to uncheck them.
* They seem to be saved in `/home/poweruser/eclipse/eclipse0/p2/org.eclipse.equinox.p2.engine/.settings/org.eclipse.equinox.p2.metadata.repository.prefs`. Setting `false` might be a solution for settings such as:

```
repositories/https\:__download.eclipse.org_releases_latest/enabled=true
repositories/https\:__download.eclipse.org_releases_latest/nickname=Latest Eclipse Simultaneous Release
repositories/https\:__download.eclipse.org_releases_latest/uri=https\://download.eclipse.org/releases/latest
```
